core/arch/arm/tee/entry_std.c

```diff
@@ -152,122 +160,174 @@ static void copy_out_param(struct tee_ta_param *ta_param, uint32_t num_params,
+				     (void *)&params[1].u.value);
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
 
-	(*num_meta) += 2;
-	return true;
+	*num_meta = 2;
+	return TEE_SUCCESS;
 }
 
-static void entry_open_session(struct thread_smc_args *smc_args,
-			struct optee_msg_arg *arg, uint32_t num_params)
+static void inject_entropy_with_timestamp(void)
 {
-	struct tee_dispatch_open_session_in in;
-	struct tee_dispatch_open_session_out out;
-	struct optee_msg_param *params = OPTEE_MSG_GET_PARAMS(arg);
-	size_t num_meta = 0;
-
-	if (!get_open_session_meta(arg, num_params, &num_meta, &in.uuid,
-				   &in.clnt_id))
-		goto bad_params;
-
-	if (!copy_in_params(params + num_meta, num_params - num_meta,
-			    &in.param))
-		goto bad_params;
-
-	(void)tee_dispatch_open_session(&in, &out);
+	TEE_Time current;
 
-	copy_out_param(&out.param, num_params - num_meta, params + num_meta);
-
-	arg->session = (vaddr_t)out.sess;
-	arg->ret = out.msg.res;
-	arg->ret_origin = out.msg.err;
-	smc_args->a0 = OPTEE_SMC_RETURN_OK;
-	return;
+	if (tee_time_get_sys_time(&current) == TEE_SUCCESS)
+		tee_prng_add_entropy((uint8_t *)&current, sizeof(current));
+}
 
-bad_params:
-	DMSG("Bad params");
-	arg->ret = TEE_ERROR_BAD_PARAMETERS;
-	arg->ret_origin = TEE_ORIGIN_TEE;
+static void entry_open_session(struct thread_smc_args *smc_args,
+			       struct optee_msg_arg *arg, uint32_t num_params)
+{
+	TEE_Result res;
+	struct optee_msg_param *params = OPTEE_MSG_GET_PARAMS(arg);
+	TEE_ErrorOrigin err_orig = TEE_ORIGIN_TEE;
+	struct tee_ta_session *s = NULL;
+	TEE_Identity clnt_id;
+	TEE_UUID uuid;
+	struct tee_ta_param param;
+	size_t num_meta;
+
+	res = get_open_session_meta(num_params, params, &num_meta, &uuid,
+				    &clnt_id);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	res = copy_in_params(params + num_meta, num_params - num_meta, &param);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	res = tee_ta_open_session(&err_orig, &s, &tee_open_sessions, &uuid,
+				  &clnt_id, TEE_TIMEOUT_INFINITE, &param);
+	if (res != TEE_SUCCESS)
+		s = NULL;
+	copy_out_param(&param, num_params - num_meta, params + num_meta);
+
+	/*
+	 * The occurrence of open/close session command is usually
+	 * un-predictable, using this property to increase randomness
+	 * of prng
+	 */
+	inject_entropy_with_timestamp();
+
+out:
+	if (s)
+		arg->session = tee_svc_kaddr_to_uref(s);
+	else
+		arg->session = 0;
+	arg->ret = res;
+	arg->ret_origin = err_orig;
 	smc_args->a0 = OPTEE_SMC_RETURN_OK;
 }
 
 static void entry_close_session(struct thread_smc_args *smc_args,
 			struct optee_msg_arg *arg, uint32_t num_params)
 {
+	TEE_Result res;
+	struct tee_ta_session *s;
 
-	if (num_params == 0) {
-		struct tee_close_session_in in;
-
-		in.sess = (TEE_Session *)(uintptr_t)arg->session;
-		arg->ret = tee_dispatch_close_session(&in);
-	} else {
-		arg->ret = TEE_ERROR_BAD_PARAMETERS;
+	if (num_params) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
 	}
 
+	s = tee_svc_uref_to_kaddr(arg->session);
```

Comment:
minor: this seems related to userland (init value based in `CFG_WITH_USER_TA)` even is underlying code is applicable here. to avoid confusion, shouldn't we have a dedicated `tee_smc_nsref_to_kaddr()`.